### PR TITLE
[CBRD-22733] batch sequential keys in an index load task

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-22733

Create a batch of keys before loading to index.

  1. Collect keys into a load task until the total disk size is larger than threshold.
  2. "index_load_task_memsize" system parameter to configure task memory size threshold.
  3. A batch of keys is loaded sequentially.

This is a step toward multi-producer task. Based on results, we can choose two approaches:

  1. If results are promising, we may scan the heap and generate batches on consecutive heap pages.
  2. If results are bad, we need to randomize the index access. We'll probably have to extract all sectors from file table and then generate batches in a random order.

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree_load.c (4 incorrect indentation(s) found)
```
  4813   std::unique_ptr < index_builder_loader_task > load_task = NULL;
  4923 							  load_context));
  4928 	  /* Increment tasks started. */
  4929 	  tasks_started++;
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/storage/btree_load.c [hunk 1]
```diff
@@ -158,26 +158,38 @@ struct index_builder_loader_context
...
+  public:
+    index_builder_loader_task (const BTID * btid, const OID * class_oid, int unique_pk,
                                index_builder_loader_context & load_context);
-    
-    void set_key (const DB_VALUE * key);
-
     ~index_builder_loader_task ();
+    
+    // add key to key set and return true if task is ready for execution, false otherwise
+    bool set_key (const DB_VALUE * key, const OID& oid);
```
Comment (razvanradulescu) [inline src/storage/btree_load.c]: rename 'set_key' -> 'add_key'
if return type is an enum of 'continue_task, full_task' the code would be more readable.

File: src/storage/btree_load.c [hunk 2]
```diff
@@ -4903,15 +4917,17 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
-      index_builder_loader_task *load_task = new index_builder_loader_task (btid_int->sys_btid, &class_oids[cur_class],
-									    &cur_oid, unique_pk, load_context);
-
-      load_task->set_key (p_dbvalue);
-
-      thread_get_manager ()->push_task (ib_workpool, load_task);
-
-      /* Increment tasks started. */
-      tasks_started++;
+      if (load_task == NULL)
```
Comment (alexstanro) [inline src/storage/btree_load.c]: The code wants to collect items into index_builder_loader_task, and start a thread when the memory threshold is reached. For readability reason:
- add proper comment
- replace code like thread_get_manager ()->push_task (ib_workpool, load_task.release ()), with:
   - task = load_task.release (); assert (load_tsk.get() == NULL);) ;thread_get_manager ()->push_task (ib_workpool, task);
 Similar for load_task.reset (new index_builder_loader_task (btid_int->sys_btid, &class_oids[cur_class], unique_pk,
                                                          load_context));

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree_load.c (4 incorrect indentation(s) found)
```
  5040 					       clear_keys ();});
  5059 bool
  5060 index_builder_loader_task::has_keys ()
  5061   constconst
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Pull request merged**